### PR TITLE
*: extend skyline pruning diagnostic info and add extra warnings

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: ref #39893

### What is changed and how it works?

- Do not limit to 1 info string per operator when recording diagnostic info for skyline pruning.
- Prepare the extra warnings that are going to be recorded in the slow log.
  - Extra warnings mean warnings that were originally only generated in some special condition, like in EXPLAIN or EXPLAIN VERBOSE.
- Make the test code in `TestSlowQuery` more clear.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -1440,8 +1440,13 @@ func (s *SessionVars) IsMPPEnforced() bool {
...
 // omit the overhead of string construction when calling with false condition.
 func (s *SessionVars) RaiseWarningWhenMPPEnforced(warning string) {
-	if s.IsMPPEnforced() && s.StmtCtx.InExplainStmt {
+	if !s.IsMPPEnforced() {
+		return
+	}
+	if s.StmtCtx.InExplainStmt {
 		s.StmtCtx.AppendWarning(errors.New(warning))
+	} else {
+		s.StmtCtx.AppendExtraWarning(errors.New(warning))
```
Comment (xuyifangreeneyes) [inline sessionctx/variable/session.go:1449]: Generating diagnostic strings may bring some overhead to the compile phase. The potential overhead is worth it for enhanced diagnosability. Maybe we can test the potential overhead after  #39893 is done(check whether compile duration increases in some typical workloads).

Comment (time-and-fate) [reply sessionctx/variable/session.go:1449]: You are right.
Though usually there would not be too many such strings, I'll run a benchmark then.

**Comment by time-and-fate** [comment]
Benchmark result:

Nightly: f150d376517e126b6ca2396f423e1709
This PR: [81701ab](https://github.com/pingcap/tidb/pull/39894/commits/81701ab3c5ed3ba15c31fce158e48c2e70600735)

#### sysbench oltp_read_write

|         | TPS           | QPS  | avg latency | P99 latency | P999 latency|
| :--------: |:-------------:| :-----:| :---: | :---: | :---: |
| Nightly      | 2781 | 55407 | 3.09 | 18.5 | 31.6 |
| This PR      | 2746 | 54704 |  3.09 | 18.8 | 32.5|

#### sysbench oltp_point_select

|         | TPS           | QPS  | avg latency | P99 latency | P999 latency|
| :-------------: |:-------------:| :-----:| :---: | :---: | :---: |
| Nightly      | 6.62 | 127890 | 1.20 | 3.91 | 11.8 |
| This PR      | 6.63 | 127950 |  1.19 | 3.91 | 11.1|

#### tpcc


|         | TPS           | QPS  | avg latency | P99 latency | P999 latency|
| :-------------: |:-------------:| :-----:| :---: | :---: | :---: |
| Nightly      | 2527 | 39238 | 4.83 | 72.9 | 124 |
| This PR      | 2565 | 39837 |  4.75 | 68.4 | 122|

**Review by qw4990 — APPROVED** [review]

**Pull request merged**